### PR TITLE
Added probing for URL + input based on protocol

### LINKED ISSUE #2253: nuclei X httpx integration
For running **http** protocol-based nuclei templates, nuclei expect URLs as input and do not work if non-probbed subdomains or IPs are provided.

Generally, we advise users to run [httpx](https://github.com/projectdiscovery/httpx) on the subdomain or IP list before passing to nuclei which is okay but has some drawbacks as well.

1. httpx will filter out all **non-http** services that could have been used with other protocols to run **non-http** templates.
2. not all the users realize or are able to debug the problems they are experiencing with input.

To improve user experience and adaptation, httpx could be integrated into nuclei to perform necessary probing to run **http based** templates.

Since this integration is supposed to be **automatic** and done internally, this allows nuclei engine to control when and what input needs to be probed depending on template input, i.e

1. Probing will be done only for the input that does not contain URL scheme
2. Probing will be done only when running HTTP based template.
3. Probbed data will be used only with HTTP-based templates.
4. Enables **uniform input for the nuclei engine** to run all the templates.

This integration is also a requirement for another nuclei integration with [uncover](https://github.com/projectdiscovery/uncover) in https://github.com/projectdiscovery/nuclei/issues/2254

An example showcasing difference between both ways to run nuclei:

### Running nuclei with httpx in between:

```
echo ger-ned380kv.nl | httpx -silent | nuclei -pt dns,http

[INF] No results found. Better luck next time!
```

### Running nuclei httpx integration:

```
echo ger-ned380kv.nl | nuclei  -pt dns,http

[servfail-refused-hosts] [dns] [info] ger-ned380kv.nl

### PR DESCRIPTION
**Pull request by Ice3man543**
## Proposed changes

Closes #2253 & #2514

## Checklist

- [ ] Pull request is created against the [dev](https://github.com/projectdiscovery/nuclei/tree/dev) branch
- [ ] All checks passed (lint, unit/integration/regression tests etc.) with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: v2/internal/runner/inputs.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+	if err != nil {
+		return nil, errors.Wrap(err, "could not get http client")
+	}
+	gologger.Verbose().Msgf("Started probing of non-http URLs")
+
+	// Probe the non-standard URLs and store them in cache
+	swg := sizedwaitgroup.New(r.options.BulkSize)
+	count := 0
+	r.hmapInputProvider.Scan(func(value string) bool {
+		if strings.Contains(value, "://") {
```
Comment (Mzack9999) [inline v2/internal/runner/inputs.go:34]: What do you think about making the test stricter by checking the prefix `https://` or `http://`, since there are many edge cases like:
- Connection strings (eg. `sql://`, `ftp://`, etc)
- URLs (eg. `host:port?param=https://xxxx`)

File: v2/internal/runner/inputs.go [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+	httpSchemes  = []string{"https", "http"}
+)
+
+// probeURL probes the scheme for a URL. first HTTPS is tried
+// and if any errors occur http is tried. If none succeeds, probing
+// is abandoned for such URLs.
+func probeURL(input string, httpclient *retryablehttp.Client) string {
+	for _, scheme := range httpSchemes {
+		formedURL := fmt.Sprintf("%s://%s", scheme, input)
+		resp, err := httpclient.Get(formedURL)
```
Comment (Mzack9999) [inline v2/internal/runner/inputs.go:66]: Do you think using `httpclient.Head` and avoid body processing could bring any benefit?

**Review by Mzack9999 — CHANGES_REQUESTED** [review]
lgtm - Suggesting a couple of possible optimizations

File: v2/internal/runner/inputs.go (hunk 2 above)
Comment (Ice3man543) [reply v2/internal/runner/inputs.go:66]: i think we should keep GET as sometimes HEAD may be disabled as security measure so we should just use the most standard way possible for probing

File: v2/internal/runner/inputs.go (hunk 1 above)
Comment (Ice3man543) [reply v2/internal/runner/inputs.go]: done

**Review by ehsandeep — CHANGES_REQUESTED** [review]
Time difference between built-in and httpx probbing.

```console
time chaos -d hackerone.com | ./nuclei -id tech-detect

[tech-detect:jsdelivr] [http] [info] https://docs.hackerone.com
[tech-detect:react] [http] [info] https://docs.hackerone.com
[tech-detect:gatsby] [http] [info] https://docs.hackerone.com
[tech-detect:jsdelivr] [http] [info] https://api.hackerone.com
[tech-detect:cloudflare] [http] [info] https://api.hackerone.com
[tech-detect:jsdelivr] [http] [info] https://www.hackerone.com
[tech-detect:google-tag-manager] [http] [info] https://www.hackerone.com
[tech-detect:cloudflare] [http] [info] https://www.hackerone.com
[tech-detect:cloudflare] [http] [info] https://support.hackerone.com/hc/en-us
[tech-detect:cloudflare] [http] [info] https://support.hackerone.com/hc
[tech-detect:cloudflare] [http] [info] https://support.hackerone.com
[tech-detect:nginx] [http] [info] https://gslink.hackerone.com

real	0m45.368s
user	0m1.291s
sys	0m0.343s
```

```console
time chaos -d hackerone.com | httpx | ./nuclei -id tech-detect

[tech-detect:react] [http] [info] https://docs.hackerone.com
[tech-detect:gatsby] [http] [info] https://docs.hackerone.com
[tech-detect:jsdelivr] [http] [info] https://docs.hackerone.com
[tech-detect:cloudflare] [http] [info] https://www.hackerone.com
[tech-detect:jsdelivr] [http] [info] https://www.hackerone.com
[tech-detect:google-tag-manager] [http] [info] https://www.hackerone.com
[tech-detect:jsdelivr] [http] [info] https://api.hackerone.com
[tech-detect:cloudflare] [http] [info] https://api.hackerone.com
[tech-detect:nginx] [http] [info] https://gslink.hackerone.com
[tech-detect:cloudflare] [http] [info] https://support.hackerone.com/hc/en-us
[tech-detect:cloudflare] [http] [info] https://support.hackerone.com/hc
[tech-detect:cloudflare] [http] [info] https://support.hackerone.com

real	0m20.840s
user	0m1.113s
sys	0m0.286s
```

**Comment by Ice3man543** [comment]
@ehsandeep this is happening due to lower retries and timeout in httpx. Setting that manually as same in nuclei does the same time for scan but results in lower count hence it is advised to keep nuclei defaults

File: v2/internal/runner/inputs.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+		if strings.HasPrefix(value, "http://") || strings.HasPrefix(value, "https://") {
+			return true
+		}
+
+		swg.Add()
+		go func(input string) {
+			defer swg.Done()
+
+			if result := probeURL(input, httpclient); result != "" {
+				count++
```
Comment (Mzack9999) [inline v2/internal/runner/inputs.go:49]: Not sure if this is thread-safe, what about using `atomic.AddIntxxx`?

**Review by Mzack9999 — CHANGES_REQUESTED** [review]
lgtm - Suggesting improvement to prevent potential data race

File: v2/internal/runner/inputs.go (hunk 3 above)
Comment (Ice3man543) [reply v2/internal/runner/inputs.go]: done

**Review by Mzack9999 — APPROVED** [review]
lgtm - let's update docs if necessary

**Review by ehsandeep — CHANGES_REQUESTED** [review]
- [x] build failing against dev branch
- [ ] Template clustering information improvement (disable when not used)
- [ ] Loaded template information enhancement

```console
[INF] Templates loaded for scan: 100 [HTTP: 44] [DNS: 43] [SSL: 20] [TCP: 30]
```
